mne/io/base.py

```diff
@@ -1242,10 +1254,48 @@ def filter(self, l_freq, h_freq, picks=None, filter_length='auto',
                 logger.info('Filtering a subset of channels. The highpass and '
                             'lowpass values in the measurement info will not '
                             'be updated.')
-        filter_data(self._data, self.info['sfreq'], l_freq, h_freq, picks,
-                    filter_length, l_trans_bandwidth, h_trans_bandwidth,
-                    n_jobs, method, iir_params, copy=False, phase=phase,
-                    fir_window=fir_window, fir_design=fir_design)
+        # Deal with annotations
+        annot = self.annotations
+        if skip_by_annotation is None:
+            if annot is not None and any(desc.upper().startswith('EDGE')
+                                         for desc in annot.description):
+                warn('skip_by_annotation defaults to [] in 0.15 but will '
+                     'change to "edge" in 0.16, set it explicitly to avoid '
+                     'this warning', DeprecationWarning)
+            skip_by_annotation = []
+        if not isinstance(skip_by_annotation, (string_types, list, tuple)):
+            raise TypeError('skip_by_annotation must be str, list, or tuple, '
+                            'got %s' % (type(skip_by_annotation)))
+        elif isinstance(skip_by_annotation, string_types):
+            skip_by_annotation = [skip_by_annotation]
+        elif not all(isinstance(skip, string_types)
+                     for skip in skip_by_annotation):
+            raise TypeError('All entries in skip_by_annotation must be str')
+        if annot is not None:
+            skips = [idx for idx, desc in enumerate(annot.description)
+                     if any(desc.upper().startswith(skip.upper())
+                            for skip in skip_by_annotation)]
+            onsets = annot.onset[skips]
+            onsets = _sync_onset(self, onsets)
+            ends = onsets + self.annotations.duration[skips]
+            del skips
+        else:
+            onsets = ends = np.array([], float)
+        order = np.argsort(onsets)
+        onsets = self.time_as_index(onsets[order])
+        ends = self.time_as_index(ends[order])
+        if len(onsets) == 0 or onsets[0] != 0:
+            onsets = np.concatenate([[0], onsets])
+            ends = np.concatenate([[0], ends])
+        if len(ends) == 1 or ends[-1] != len(self.times):
+            onsets = np.concatenate([onsets, [len(self.times)]])
+            ends = np.concatenate([ends, [len(self.times)]])
```

Comment:
I checked and there is *some* duplication but it's not identical. I'll move it and look further to see if I can refactor it, though.